frontend/src/metabase/core/components/TabButton/TabButton.tsx

```diff
@@ -155,26 +151,30 @@ const TabButton = forwardRef(function TabButton(
   );
 });
 
-export interface RenameableTabButtonProps
-  extends Omit<TabButtonProps, "onEdit" | "onFinishEditing" | "isEditing"> {
+export interface RenameableTabButtonProps<T>
+  extends Omit<TabButtonProps<T>, "onEdit" | "onFinishEditing" | "isEditing"> {
   onRename: (newLabel: string) => void;
   renameMenuLabel?: string;
   renameMenuIndex?: number;
 }
 
-export function RenameableTabButton({
-  label: originalLabel,
+export function RenameableTabButton<T>({
+  label: labelProp,
   menuItems: originalMenuItems = [],
   onRename,
   renameMenuLabel = t`Rename`,
   renameMenuIndex = 0,
   ...props
-}: RenameableTabButtonProps) {
-  const [label, setLabel] = useState(originalLabel);
+}: RenameableTabButtonProps<T>) {
+  const [label, setLabel] = useState(labelProp);
   const [prevLabel, setPrevLabel] = useState(label);
   const [isEditing, setIsEditing] = useState(false);
   const inputRef = useRef<HTMLInputElement>(null);
 
+  useEffect(() => {
+    setLabel(labelProp);
```

Comment:
If the label prop changes, override this component's internal `label` state with the new prop value.

This is needed to make renaming work on dashboards. Renaming will dispatch a redux action that causes the tab's name to change in the store. That new name is then passed as a prop, so we need this `useEffect` to ensure it updates.